Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -164,6 +244,125 @@ theorem testBit_two_pow (n m : ℕ) : testBit (2 ^ n) m = (n = m) := by
+This is an alternative to using `List` altogether. It will be used for bitadd, bitneg, bitmul etc.-/
+def toNat (f : Nat → Bool) (z : Nat) : Nat → Nat
+  | 0 => z.bit (f 0)
+  | i + 1 => toNat f (z.bit (f (i + 1))) i
+
+theorem toNat_succ : toNat f z i = 2^(i + 1) * z + toNat f 0 i := by
+  induction' i with i ih generalizing z
+  · simp [toNat, bit_val]
+  · simp only [toNat, @ih (bit (f (i + 1)) 0), @ih (bit (f (i + 1)) z)]
+    rw [bit_val, mul_add, ← mul_assoc, ← pow_succ]
+    simp [bit_val, add_assoc] 
+
+theorem toNat_lt : toNat f 0 i < 2^(i + 1) := by
+  induction' i with i ih
+  · simp [toNat, bit_val, lt_succ, Bool.toNat_le_one]
+  · simp only [toNat]
+    rw [toNat_succ]
+    cases' (f (i + 1)) <;> simp [ih, two_pow_succ] at * <;> linarith
+
+/-- The `ith` bit of `toNat` is the function at `i`. 
+This is used extensively in the proof of each of the bitadd, bitneg, bitmul etc.-/
+theorem toNat_testBit (h1: i ≤ j): (toNat f 0 j).testBit i = f i := by
+  induction' j with j ih generalizing i
+  · simp [nonpos_iff_eq_zero.mp h1, toNat, bit_val]
+  · cases' lt_or_eq_of_le h1 with h1 h1
+    · rw [← ih (show i ≤ j by linarith), toNat, toNat_succ, testBit_translate h1]
+    · rw [h1, toNat, toNat_succ, bit_toNat, testBit_translate' (toNat_lt)]
+
+
+/-! ### Unsigned Less Than-/
+
+/-- Unsigned less than criterion (`n < m`) for natural numbers `n`, `m`
+and `w` - the length of their binary representation.-/
+def bitult (n m w : Nat) := loop n m (w - 1) 
+where
+  loop (n m : Nat) : Nat →  Prop
+    | 0     => ¬ n.testBit 0  ∧ m.testBit 0
+    | i + 1 => (¬ n.testBit (i + 1) ∧ m.testBit (i + 1)) 
+               ∨ (¬ (n.testBit (i + 1) ∧ ¬ m.testBit (i + 1)) ∧ loop n m i)
+
+theorem bitult_of_ult (hm: m < 2^w) (h1: n < m) : bitult n m w := by
+  have ⟨i, hi1, hi2, hi3⟩ := testBit_of_lt h1
+  suffices goal: ∀ j, i + 1 ≤ j → bitult n m j from goal w (lt_of_testBit_true hi2 hm)
+  apply le_induction
+  · cases' i <;> simp [bitult, bitult.loop, hi1, hi2]
+  · intros j hj ih
+    have ⟨j', hj'⟩ := exists_eq_add_of_le' (le_of_add_le_right hj)
+    simp only [bitult, bitult.loop, hj', succ_sub_one j'] at ih ⊢ 
+    simp [ih, hi3 (j' + 1) (by linarith)]
+
+
+/-! ### Addition-/
+
+/-- Carry function at the `i`th bit for binary addition on `n` and `m`.-/
+def bitcarry (n m : Nat) : Nat → Bool
+  | 0     => false
+  | i + 1 => (n.testBit i && m.testBit i) || ((n.testBit i ^^ m.testBit i) && bitcarry n m i)
+
+/-- Binary addition is `toNat` specialized with the right function on the bits `f`.-/
+@[simp] def bitadd (n m i : Nat) := 
+  toNat (λ j => (n.testBit j ^^ m.testBit j) ^^ bitcarry n m j) 0 i
+
+lemma unfold_carry (n m i : Nat) : (bitcarry n m (i + 1)).toNat = 
+  ((Nat.testBit n i && Nat.testBit m i) || 
+  ((Nat.testBit n i ^^ Nat.testBit m i) && bitcarry n m i)).toNat := by 
+  simp [bitcarry]
+
+lemma bitadd_eq_add_base : n % (2^(i + 1)) + m % (2^(i + 1)) = 
```

Comment:
This should have a better name. `bitadd` is not on the LHS of the Type so it shouldn't be on the LHS in the name. Something like maybe switching the LHS and RHS in the statement and calling it `bitadd_add_two_pow_mul_bitcarry`